server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,542 @@
+	clusterSize, err := d.getClusterSize()
+	if err != nil {
+		if err == ErrSizeNotFound || err == ErrBadSizeKey {
+			return nil, 0, 0, err
+		}
+
+		return d.checkClusterRetry()
+	}
+
+	members, rev, err := d.getClusterMembers()
+	if err != nil {
+		return d.checkClusterRetry()
+	}
+
+	// find self position
+	memberSelfId := getMemberKey(d.cluster, d.id.String())
+	idx := 0
+	for k := range members {
+		if k == memberSelfId {
+			break
+		}
+		if idx >= clusterSize-1 {
+			return members, clusterSize, rev, ErrFullCluster
+		}
+		idx++
+	}
+	return members, clusterSize, rev, nil
+}
+
+func (d *discovery) registerSelfRetry(contents string) error {
+	if d.retries < nRetries {
+		d.logAndBackoffForRetry("register member itself")
+		return d.registerSelf(contents)
+	}
+	return ErrTooManyRetries
+}
+
+func (d *discovery) registerSelf(contents string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.cfg.RequestTimeOut)
+	memberKey := getMemberKey(d.cluster, d.id.String())
+	_, err := d.c.Put(ctx, memberKey, contents)
+	cancel()
+
+	if err != nil {
+		d.lg.Warn(
+			"failed to register members itself to the discovery service",
+			zap.String("discovery-url", d.durl),
+			zap.String("memberKey", memberKey),
+			zap.Error(err),
+		)
+		return d.registerSelfRetry(contents)
+	}
+
+	d.lg.Info(
+		"register member itself successfully",
+		zap.String("discovery-url", d.durl),
+		zap.String("memberKey", memberKey),
+		zap.String("memberInfo", contents),
+	)
+
+	return nil
+}
+
+func (d *discovery) waitNodes(members map[string]memberInfo, clusterSize int, rev int64) (map[string]memberInfo, error) {
+	// watch from the next revision
+	membersKeyPrefix := getMemberKeyPrefix(d.cluster)
+	w := d.c.Watch(context.Background(), membersKeyPrefix, clientv3.WithPrefix(), clientv3.WithRev(rev+1))
+
+	d.lg.Info(
+		"waiting for peers from discovery service",
+		zap.String("discovery-url", d.durl),
+		zap.Int("clusterSize", clusterSize),
+		zap.Int("found-peers", len(members)),
+	)
+
+	// waiting for peers until all needed peers are returned
+	for wresp := range w {
+		for _, ev := range wresp.Events {
+			mKey := strings.TrimSpace(string(ev.Kv.Key))
+			mValue := strings.TrimSpace(string(ev.Kv.Value))
+
+			if d.shouldIgnorePeer(members, mKey, mValue) {
+				continue
+			}
+
+			d.lg.Info(
+				"found peer from discovery service",
+				zap.String("discovery-url", d.durl),
+				zap.String("memberKey", mKey),
+				zap.String("memberInfo", mValue),
+			)
+
+			members[mKey] = memberInfo{peerURL: mValue, createRev: ev.Kv.CreateRevision}
+		}
+
+		if len(members) >= clusterSize {
+			break
+		}
+	}
+
+	d.lg.Info(
+		"found all needed peers from discovery service",
+		zap.String("discovery-url", d.durl),
+		zap.Int("clusterSize", clusterSize),
+		zap.Int("found-peers", len(members)),
+	)
+	return members, nil
+}
+
+func (d *discovery) shouldIgnorePeer(members map[string]memberInfo, memberKey, memberInfo string) bool {
```

Comment:
Note about the code readability. I noticed that there are a lot of methods on `(d *discovery)` that only minimally access `*discovery` and mostly operate on `members map[string]memberInfo`. For example this function `shouldIngorePeer`, here we only need membersKeyPrefix and logger, most of operation is done on the `members`. 

This is example of breaking https://en.wikipedia.org/wiki/Law_of_Demeter where struct `discovery` know too much on how members. This is a nit, however code would be more readable if we created a struct or type alias for `map[string]memberInfo` with those functions.

Reply:
Fixed.